rclcpp/include/rclcpp/node_options.hpp

```diff
@@ -256,6 +271,8 @@ class NodeOptions
 
   rmw_qos_profile_t parameter_event_qos_profile_ {rmw_qos_profile_parameter_events};
 
+  bool allow_undeclared_parameters_ {false};
```

Comment:
Disallowing undeclared parameters by default is going to make a lot of people unhappy when they switch to Dashing, since they have had no chance to change to the "declared parameters" world before this.  It seems like we should allow undeclared parameters by default in Dashing (possibly with a warning), then maybe switch this to `false` in E-turtle.

Reply:
I'd be willing to do that, though it will make all of this more complicated because this goes from a bool to a choice (allow, disallow, allow but do deprecation warning).

On the flip side, they can simply create their nodes in dashing with a different value for this option, so the impact is relatively small. Or they can do more work, i.e. declare their parameters, which has more long term benefits.

Anyone else have an opinion on this?

Reply:
> On the flip side, they can simply create their nodes in dashing with a different value for this option, so the impact is relatively small. Or they can do more work, i.e. declare their parameters, which has more long term benefits.

True.  My point of view is coming from code that I'm going to need to port from Crystal to Dashing, and with this change I know things won't work out of the box.  I think there are 2 questions:

1.  What do we *eventually* want the default state of `allow_undeclared_parameters` to be?
1.  How do we get there?

The first question basically comes down to flexibility vs. reliability.  Right now (and in ROS 1) we're basically on the side of flexibility, but switching this to false gets us closer to reliability.  I'm OK with that, but it is a change from what we've traditionally done.

For the second question we should consider whether we want to start tick-tocking for a change like this, or whether we should just add it to the porting guide.

Putting it that way, I'd be fine with either tick-tocking with a warning in Dashing and switching to false in E-Turtle, or changing this to false and adding some notes to the Dashing porting guide.

That being said, I'd like to hear other opinions on this change.

Reply:
I would argue that the default of `allow_undeclared_parameters` should be `true` since that matches past behavior (in ROS 2 as well as ROS 1). To consider the opposite a clear argument should be made why that should be the preferred default.

Reply:
Whether or not `allow_undeclared_parameters` is the default, I think any node with a fixed set of parameters should set it to `false` and declare them.

If undeclared parameters are allowed then misspellings of parameters can only be caught if the node author explicitly checks them. If only declared parameters are allowed, the node author still has to check them after declaring their parameters, but they could use a common function that asserts no undeclared parameters remain in the initial parameter list.

If a user interacting with a node is partially trusted, that is they may change parameters but should not be able to bring the system down then undeclared parameters are a potential attack. A malicious user could set many parameters until the node runs out of memory. A node author can't know if someone running their node is going to have a use case with partially trusted users, so if allowing undeclared parameters is the default then someone with this use case must modify existing nodes to declare their parameters.

Reply:
I basically agree with @sloretz, it's safer against both typo's (misspelled parameter names) and denial of service attacks (intentional or accidental). It also has implications for safety, in that if you only allow declared parameters, you're unlikely to use more memory than you expect or something like that (declared parameters can be pre-allocated at some level).

Reply:
That's the argument for making it true at some point, separately, I'm fine with doing a tick-tock roll out if people are concerned about disruption to Crystal users that are upgrading. I defaulted to just changing it because we hadn't been doing backwards compatible changes in other cases (like the `NodeOptions` pr). I figured documenting it in the release notes would be sufficient.

Reply:
Summarizing from ROS 2 meeting:

- long term we do want to require parameters to be declared
- we will not do the tick-tock, because it would require an immediately deprecated option (allow undeclared but warn) and because the burden of setting one flag in the `NodeOptions` is small if migrating users do not want to deal with declaring parameters

I will rework the tests and documentation to be clearer about this, and then I will draft a pr to the dashing migration notes that is as thorough as possible.